### PR TITLE
[DROOLS-5353] Avoid ClassLoader retention in LambdaIntrospector

### PR DESCRIPTION
**Pull request by tkobayas**
In order to avoid ClassLoader retention, this PR changes the cache to use WeakHashMap where a ClassLoader is a weak key. So if the ClassLoader is no longer used, the entry will be removed (= WeakHashMap doesn't prevent ClassLoader from being garbage-collected).

ClassIdentifier is split into ClassLoader and className so the Map structure is also split. In the WeakHashMap, methodFingerprintsMap is put as a value per className. So the methodFingerprintsMap itself isn't changed by this PR.

Now ClassLoader wouldn't leak regardless of externaliseCanonicalModelLambda is enabled or not. Btw, the cache size doesn't matter much now. Even if externaliseCanonicalModelLambda is enabled, the default value '32' wouldn't be a problem because ClassLoarders will be garbage-collected.

### CODE REVIEW

**Review by lucamolteni — APPROVED** [review]

**Review by hellowdan — APPROVED** [review]

File: drools-model/drools-model-compiler/src/test/java/org/drools/modelcompiler/util/LambdaIntrospectorTest.java [hunk 1]
```diff
@@ -45,21 +45,23 @@ public boolean test(org.drools.modelcompiler.domain.Person p) {
...
     public void testMethodFingerprintsMapCacheSize() throws Exception {
-        // Because methodFingerprintsMap is static, this property can be testable when you run this test method only
+        // To test system property, you need to run this test method only.
         // (mvn test -Dtest=LambdaIntrospectorTest#testMethodFingerprintsMapCacheSize)
         // System.setProperty(LambdaIntrospector.LAMBDA_INTROSPECTOR_CACHE_SIZE, "0");
 
         LambdaIntrospector lambdaIntrospector = new LambdaIntrospector();
 
-        Field field = LambdaIntrospector.class.getDeclaredField("methodFingerprintsMap");
+        Field field = LambdaIntrospector.class.getDeclaredField("methodFingerprintsMapPerClassLoader");
```
Comment (mariofusco) [inline drools-model/drools-model-compiler/src/test/java/org/drools/modelcompiler/util/LambdaIntrospectorTest.java]: I don't think it is a good idea to read the cache field via reflection in this way, even in a test. Please add a package private accessor method and use it instead.

Comment (tkobayas) [reply drools-model/drools-model-compiler/src/test/java/org/drools/modelcompiler/util/LambdaIntrospectorTest.java]: Sure! Added a package private accessor so got rid of reflection.

**Comment by tkobayas** [comment]
The failure in jbpm-human-task-audit was caused by Enforcer so not related to this PR.

```
02:55:59  [ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project jbpm-human-task-audit: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
02:55:59  org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project jbpm-human-task-audit: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
02:55:59      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
02:55:59      at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:200)
02:55:59      at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:196)
02:55:59      at java.util.concurrent.FutureTask.run (FutureTask.java:266)
02:55:59      at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
02:55:59      at java.util.concurrent.FutureTask.run (FutureTask.java:266)
02:55:59      at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
02:55:59      at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
02:55:59      at java.lang.Thread.run (Thread.java:748)
02:55:59  Caused by: org.apache.maven.plugin.MojoExecutionException: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
02:55:59      at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:243)
02:55:59      at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
02:55:59      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
02:55:59      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
02:55:59      at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:200)
02:55:59      at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:196)
02:55:59      at java.util.concurrent.FutureTask.run (FutureTask.java:266)
02:55:59      at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
02:55:59      at java.util.concurrent.FutureTask.run (FutureTask.java:266)
02:55:59      at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
02:55:59      at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
02:55:59      at java.lang.Thread.run (Thread.java:748)
```

**Comment by mariofusco** [comment]
jenkins test

**Comment by tkobayas** [comment]
Now no longer Enforcer error for jbpm-human-task-audit but ArtifactResolverException for optaweb-employee-rostering-frontend:zip

**Comment by tkobayas** [comment]
jenkins please test again

**Pull request merged**